File: c/src/reverse_words_in_a_string.c
#include "reverse_words_in_a_string.h"

#include <string.h>

static void reverse(char *s, size_t start, size_t end) {
    const size_t sum = start + end, half = sum / 2;
    char tmp;
    for (size_t i = start; i < half; ++i) {
        tmp = s[i];
        s[i] = s[sum - i - 1];
        s[sum - i - 1] = tmp;
    }
}

void reverseWords_151_1(char *s) {
    if (s == NULL) return;
    const size_t len = strlen(s);
    if (len == 0) return;
    reverse(s, 0, len);

    size_t i, j, size = 0;
    for (i = 0; i < len && s[i] == ' '; ++i);
    while (i < len) {
        for (j = i + 1; j < len && s[j] != ' '; ++j);
        reverse(s, i, j);
        memmove(s + size, s + i, j - i);
        size += j - i;
        if (size < len) s[size++] = ' ';
        for (i = j + 1; i < len && s[i] == ' '; ++i);
    }
    if (size == 0) s[0] = '\0';
    else if (s[size - 1] == ' ') s[size - 1] = '\0';
}

File: java/src/main/java/io/binac/leetcode/ReverseWordsInAString.java
package io.binac.leetcode;

/**
 * Given an input string, reverse the string word by word.
 * <p>
 * <p>Example:
 * <blockquote><pre>
 *     Input: "the sky is blue",
 *     Output: "blue is sky the".
 * </blockquote></pre>
 * Note:
 * <p>
 * <p>A word is defined as a sequence of non-space characters.
 * <p>Input string may contain leading or trailing spaces. However, your reversed string should not contain leading or trailing spaces.
 * <p>You need to reduce multiple spaces between two words to a single space in the reversed string.
 * <p>
 * <p>Follow up: For C programmers, try to solve it in-place in O(1) space.
 */
public class ReverseWordsInAString {
    public static class Solution1 {
        public String reverseWords(String s) {
            String words[] = s.trim().split("\\s+");
            final int half = words.length / 2;
            for (int i = 0; i < half; ++i) {
                String tmp = words[i];
                words[i] = words[words.length - i - 1];
                words[words.length - i - 1] = tmp;
            }
            return String.join(" ", words);
        }
    }
}
